I think the updated version is missing a \n or something.  The generated asm looks like this:

```
        lgr     %r2, %r5
        .loc    1       32
brasl   %r14, caml_raise_exn@PLT.L138:
.L136:  brasl   %r14, caml_call_gc@PLT
.L137:  brcl    15, .L135
        .globl  camlPervasives__invalid_arg_1007
```

and that causes errors like:

```
/tmp/camlasmc1decc.s: Assembler messages:
/tmp/camlasmc1decc.s:408: Error: junk at end of line: `.L138:'
/tmp/camlasmc1decc.s:435: Error: junk at end of line: `.L143:'
/tmp/camlasmc1decc.s:1286: Error: junk at end of line: `.L251:'
/tmp/camlasmc1decc.s:1476: Error: junk at end of line: `.L284:'
File "/home/rjones/d/fedora/fedora-ocaml/stdlib/pervasives.ml", line 1:
Error: Assembler error, input left in file /tmp/camlasmc1decc.s
Makefile.shared:105: recipe for target 'pervasives.cmx' failed
```

--- a/Changes
+++ b/Changes
@@ -77,6 +77,9 @@ Next version (4.05.0):
 
 ### Bug fixes
 
+- PR#7405, GPR#903: s390x: Fix address of caml_raise_exn in native dynlink modules
+  (Richard Jones, review by Xavier Leroy)
+
 - GPR#795: remove 256-character limitation on Sys.executable_name
   (Xavier Leroy)
 

--- a/asmcomp/s390x/emit.mlp
+++ b/asmcomp/s390x/emit.mlp
@@ -611,7 +611,7 @@ let emit_instr i =
     | Lraise k ->
         begin match k with
         | Cmm.Raise_withtrace ->
-          `	brasl	%r14, {emit_symbol "caml_raise_exn"}\n`;
+          `	{emit_call "caml_raise_exn"}\n`;
           let lbl = record_frame Reg.Set.empty true i.dbg in
           `{emit_label lbl}:\n`
         | Cmm.Raise_notrace ->

